x-pack/filebeat/input/o365audit/input.go

```diff
@@ -0,0 +1,302 @@
+			msg := common.MapStr{}
+			msg.Put("error.message", err.Error())
+			event := beat.Event{
+				Timestamp: time.Now(),
+				Fields:    msg,
+			}
+			inp.reportEvent(event)
+		}
+	}()
+}
+
+func (inp *o365input) reportEvent(event beat.Event) bool {
+	return inp.outlet.OnEvent(event)
+}
+
+// Stop terminates the o365 input.
+func (inp *o365input) Stop() {
+	inp.log.Info("Stopping input " + inputName)
+	defer inp.log.Info(inputName + " stopped.")
+	defer inp.outlet.Close()
+	inp.cancel()
+}
+
+// Wait terminates the o365input and waits for all the pollers to finalize.
+func (inp *o365input) Wait() {
+	inp.Stop()
+	inp.wg.Wait()
+}
+
+func (inp *o365input) loadLastLocation(key stream) cursor {
+	period := inp.config.API.MaxRetention
+	retentionLimit := time.Now().UTC().Add(-period)
+	cursor, err := inp.storage.Load(key)
+	if err != nil {
+		if err == errStateNotFound {
+			inp.log.Infof("No saved state found. Will fetch events for the last %v.", period.String())
+		} else {
+			inp.log.Errorw("Error loading saved state. Will fetch all retained events. "+
+				"Depending on max_retention, this can cause event loss or duplication.",
+				"error", err,
+				"max_retention", period.String())
+		}
+		cursor.timestamp = retentionLimit
+	}
+	if cursor.timestamp.Before(retentionLimit) {
+		inp.log.Warnw("Last update exceeds the retention limit. "+
+			"Probably some events have been lost.",
+			"resume_since", cursor,
+			"retention_limit", retentionLimit,
+			"max_retention", period.String())
+		// Due to API limitations, it's necessary to perform a query for each
+		// day. These avoids performing a lot of queries that will return empty
+		// when the input hasn't run in a long time.
+		cursor.timestamp = retentionLimit
+	}
+	return cursor
+}
+
+var errTerminated = errors.New("terminated due to output closed")
+
+// Report returns an action that produces a beat.Event from the given object.
+func (env apiEnvironment) Report(doc common.MapStr, private interface{}) poll.Action {
+	return func(poll.Enqueuer) error {
+		if !env.Callback(toBeatEvent(doc, private)) {
+			return errTerminated
+		}
+		return nil
+	}
+}
+
+// ReportAPIError returns an action that produces a beat.Event from an API error.
+func (env apiEnvironment) ReportAPIError(err apiError) poll.Action {
+	return func(poll.Enqueuer) error {
+		if !env.Callback(err.ToBeatEvent()) {
+			return errTerminated
+		}
+		return nil
+	}
+}
+
+func (env apiEnvironment) toBeatEvent(doc common.MapStr, private interface{}) beat.Event {
+	var errs multierror.Errors
+	ts, err := getDateKey(doc, "CreationTime", apiDateFormats)
+	if err != nil {
+		ts = time.Now()
+		errs = append(errs, errors.Wrap(err, "failed parsing CreationTime"))
+	}
+	b := beat.Event{
+		Timestamp: ts,
+		Fields: common.MapStr{
+			fieldsPrefix: doc,
+		},
+		Private: private,
+	}
+	if env.Config.SetIDFromAuditRecord {
+		if id, err := getString(doc, "Id"); err == nil && len(id) > 0 {
+			b.SetID(id)
```

Comment:
Any ideas on this? Sounds like a good idea to me as it will prevent duplicates, but I don't see any other data source setting the document ID.

Reply:
👍  The Google PubSub inputs sets an ID (maybe s3 does too). I think the pubsub code might need updated after #15859. I see we're using `_id` rather than `id` now.